qiskit_experiments/data_processing/data_processor.py

```diff
@@ -0,0 +1,105 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Class that ties together actions on the data."""
+
+from typing import Any, Dict, List, Tuple, Union
+
+from qiskit_experiments.data_processing.base import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class DataProcessor:
+    """
+    Defines the actions done on the measured data to bring it in a form usable
+    by the calibration analysis classes.
+    """
+
+    def __init__(self, data_actions: List[DataAction] = None):
+        """Create a chain of data processing actions.
+
+        Args:
+            data_actions: A list of data processing actions to construct this data processor with.
+                If None is given an empty DataProcessor will be created.
+        """
+        self._nodes = []
+
+        if data_actions:
+            for node in data_actions:
+                self.append(node)
+
+        self._history = []
+
+    @property
+    def history(self) -> List[Tuple[str, Dict[str, Any]]]:
+        """
+        Returns:
+            The history of the data processor. The ith tuple in the history corresponds to the
+            output of the ith node. Each tuple corresponds to (node name, data dict).
+        """
+        return self._history
+
+    def append(self, node: DataAction):
+        """
+        Append new data action node to this data processor.
+
+        Args:
+            node: A DataAction that will process the data.
+
+        Raises:
+            DataProcessorError: if the output of the last node does not match the input required
+                by the node to be appended.
+        """
+        if len(self._nodes) == 0:
+            self._nodes.append(node)
+        else:
+            if self._nodes[-1].__node_output__ not in node.node_inputs:
```

Comment:
I am not sure if the system of matching `_accepted_inputs` to `__node_output__` between each pair of steps provides the documentation / sanity checking that we want for a data processing pipe line. For example, one possible point of confusion is thinking that an analysis class is working with averaged level 1 data when it is really working with single shots. Specifying `memory_real` with the current `ToReal` `DataAction` would not help since that covers both the averaged and single-shot cases. A data analysis routine might want "one uncalibrated scalar value per measurement" but I don't know that one would specifically want the real or imaginary value of the measurement. I wonder if there is a way to propagate qualities of the data (averaged, kerneled, scalar, discriminated) forward through the processing steps without things getting too complicated. Each step could add labels, but some steps would need to remove them as well and that could be more prone to errors.

Reply:
Changed to directly returning data here a0d0903